File: backend/src/test/java/com/mycompany/online_shop_backend/controllers/OrderControllerTests.java
package com.mycompany.online_shop_backend.controllers;

import com.google.gson.Gson;
import com.mycompany.online_shop_backend.domain.*;
import com.mycompany.online_shop_backend.dto.response.CreatedOrderResponse;
import com.mycompany.online_shop_backend.dto.response.UserOrderDto;
import com.mycompany.online_shop_backend.dto.services.BookDto;
import com.mycompany.online_shop_backend.dto.request.OrderRequest;
import com.mycompany.online_shop_backend.dto.services.OrderDto;
import com.mycompany.online_shop_backend.repositories.UserRepository;
import com.mycompany.online_shop_backend.security.SecurityConfiguration;
import com.mycompany.online_shop_backend.security.TokenAuthenticationFilter;
import com.mycompany.online_shop_backend.security.TokenProperties;
import com.mycompany.online_shop_backend.security.UserDetailsServiceImpl;
import com.mycompany.online_shop_backend.services.OrderService;
import com.mycompany.online_shop_backend.services.security.SecurityService;
import com.mycompany.online_shop_backend.services.security.TokenService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.http.MediaType;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.web.servlet.MockMvc;

import javax.servlet.http.HttpServletRequest;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(OrderController.class)
@Import({TokenAuthenticationFilter.class,
        TokenProperties.class,
        TokenService.class,
        SecurityConfiguration.class,
        UserDetailsServiceImpl.class,
        UserRepository.class})
public class OrderControllerTests {
    private final User userOne = new User(
            1L,
            "Name One",
            "Surname One",
            "userOne@test.com",
            "Encoded Start01#"
    );
    private final Author authorOne = new Author(1L, "Author One");
    private final Author authorTwo = new Author(2L, "Author Two");

    private final Book bookOne = new Book(
            1L,
            "Book One",
            "Description One",
            authorOne,
            "/imageOne",
            22.95
    );
    private final Book bookTwo = new Book(
            2L,
            "Book Two",
            "Description Two",
            authorTwo,
            "/imageTwo",
            46.00
    );
    private final BookDto bookOneDto = BookDto.toDto(bookOne);
    private final BookDto bookTwoDto = BookDto.toDto(bookTwo);
    private final List<BookDto> bookDtos = List.of(bookOneDto, bookTwoDto);

    private final Order order = new Order(
            1L,
            userOne.getFirstName() + " " + userOne.getLastName(),
            new Address(1L, "Address, 1"),
            new Phone(1L, "+1111 1111"),
            userOne.getEmail(),
            LocalDateTime.now(),
            new HashSet<>()
    );
    private final OrderRequest orderRequest = new OrderRequest(
            userOne.getFirstName() + " " + userOne.getLastName(),
            "Address, 1",
            "+1111 1111",
            userOne.getEmail(),
            bookDtos
    );
    private final OrderDto orderDto = OrderDto.toDto(order);
    private final CreatedOrderResponse createdOrderResponse = CreatedOrderResponse.toDto(orderDto);
    private final UserOrderDto userOrderDto = UserOrderDto.toDto(orderDto);
    private final Set<OrderBook> orderBooks = Set.of(new OrderBook(order, bookOne), new OrderBook(order, bookTwo));

    private final Gson gson = new Gson();

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private OrderService orderService;

    @MockBean
    private UserRepository userRepository;

    @MockBean
    private SecurityService securityService;

    @BeforeEach
    void setup() {
        order.setOrderBooks(orderBooks);
    }

    @Test
    public void createOrder() throws Exception {
        when(orderService.save(any(OrderRequest.class))).thenReturn(orderDto);

        mockMvc.perform(
                post("/v1/orders")
                        .contentType(MediaType.APPLICATION_JSON)
                        .accept(MediaType.APPLICATION_JSON)
                        .content(gson.toJson(orderRequest)))
                .andExpect(status().isCreated())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(content().json(gson.toJson(createdOrderResponse)));
    }

    @WithMockUser(username = "userOne@test.com")
    @Test
    public void getUserOrders() throws Exception {
        when(securityService.getUsernameFromRequest(any(HttpServletRequest.class))).thenReturn(userOne.getEmail());
        when(orderService.getOrdersByEmail(userOne.getEmail())).thenReturn(List.of(orderDto));

        mockMvc.perform(
                get("/v1/users/{id}/orders", userOne.getId())
                        .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(content().json(gson.toJson(List.of(userOrderDto))));
    }
}

File: order-service/src/test/kotlin/com/example/orderservice/api/rest/OrderControllerTests.kt
package com.example.orderservice.api.rest

import com.example.orderservice.api.rest.model.ErrorCode
import com.example.orderservice.api.rest.model.OrderSearchRequest
import com.example.orderservice.mapping.OrderMapper.toModel
import com.example.orderservice.repository.OrderRepository
import com.example.orderservice.test.IntegrationTest
import com.example.orderservice.test.OrderTestData.orderEntity
import com.fasterxml.jackson.databind.ObjectMapper
import org.assertj.core.api.Assertions.assertThat
import org.hamcrest.core.StringContains.containsString
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import java.util.*

@IntegrationTest
class OrderControllerTests {

    @Autowired
    lateinit var mockMvc: MockMvc

    @Autowired
    lateinit var orderRepository: OrderRepository

    @Autowired
    lateinit var objectMapper: ObjectMapper

    @Test
    fun `search orders by userId`() {
        val order = orderRepository.save(orderEntity()).toModel()
        val request = OrderSearchRequest(order.userId)

        val result = mockMvc.post("/api/v1/orders/search?page=0") {
            contentType = MediaType.APPLICATION_JSON
            accept = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsString(request)
        }.andExpect {
            status { isOk() }
            content { contentType(MediaType.APPLICATION_JSON) }
        }.andReturn()

        val expectedOrder = objectMapper.writeValueAsString(order)
        assertThat(result.response.contentAsString).contains(expectedOrder)
    }

    @Test
    fun `search orders by non-existing userId`() {
        val request = OrderSearchRequest(UUID.randomUUID())

        mockMvc.post("/api/v1/orders/search?page=0") {
            contentType = MediaType.APPLICATION_JSON
            accept = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsString(request)
        }.andExpect {
            status { isOk() }
            content { contentType(MediaType.APPLICATION_JSON) }
            content { string(containsString(""""content":[]""")) }
        }
    }

    @Test
    fun `search with non-supported sorting`() {

    }

    @Test
    fun `get order by id`() {

    }

    @Test
    fun `get order by non-existing id`() {
        mockMvc.get("/api/v1/orders/${UUID.randomUUID()}") {
            accept = MediaType.APPLICATION_JSON
        }.andExpect {
            status { isNotFound() }
            content { contentType(MediaType.APPLICATION_JSON) }
            content { string(containsString(ErrorCode.RESOURCE_NOT_FOUND.name)) }
        }
    }

    @Test
    fun `create order`() {

    }

    @Test
    fun `duplicate create order request`() {

    }
}